File: src/main/java/caches/GitParser.java
package caches;

import caches.changes.*;
import caches.records.FilePointer;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.diff.DiffEntry;
import org.eclipse.jgit.lib.AbbreviatedObjectId;
import org.eclipse.jgit.lib.Constants;
import org.eclipse.jgit.lib.ObjectId;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.revwalk.RevCommit;
import org.eclipse.jgit.revwalk.RevWalk;
import org.eclipse.jgit.treewalk.TreeWalk;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static caches.GlobalVariables.*;

public class GitParser {

    private static final Logger LOG = LoggerFactory.getLogger(GitParser.class);
    private final Git git;
    private final Repository repository;
    private final List<ChangeProcessor> indexes;

    public GitParser(Git git, List<ChangeProcessor> indices) {
        this.git = git;
        repository = git.getRepository();
        this.indexes = indices;
    }

    public void parse() {
        try (RevWalk walk = new RevWalk(repository)) {
            Deque<RevCommit> commits = new ArrayDeque<>();
            {
                ObjectId head = repository.resolve(Constants.HEAD);
                walk.markStart(walk.parseCommit(head));
                walk.forEach(commits::add);
            }
            if (!commits.iterator().hasNext()) {
                throw new RuntimeException("Repository hasn't commits");
            }
            LOG.info(String.format("%d commits finded to process", commits.size()));
            System.out.printf("%d finded to process%n", commits.size());
            var firstCommit = commits.removeLast();
            parseFirstCommit(firstCommit);
            var prevCommit = firstCommit;
            while (!commits.isEmpty()) {
                if (commits.size() % 100 == 0) {
                    System.out.printf("Remaining %d commits%n", commits.size());
                }
                if (commits.size() == 386622) {
                    break;
                }
                var commit = commits.removeLast();
                parseCommit(commit, prevCommit);
                prevCommit = commit;
            }
        } catch (GitAPIException | IOException e) {
            throw new RuntimeException(e);
        }
    }



    void sendChanges(List<Change> changes) {
        revisions.getAndIncrement();
        changes.forEach(it -> {
            switch (it) {
                case FileChange fileChange -> tryRegisterNewFile(fileChange.getPlace().file());
                case FileHolderChange fileHolderChange -> {
                    tryRegisterNewFile(fileHolderChange.getOldFileName());
                    tryRegisterNewFile(fileHolderChange.getNewFileName());
                }
            }
        });
        indexes.forEach(it -> it.prepare(changes));
        currentRevision.set(revisions.get());
    }

    private void parseCommit(RevCommit commit, RevCommit prevCommit) throws IOException, GitAPIException {
        try (var tw = new TreeWalk(repository)) {
            tw.addTree(prevCommit.getTree());
            tw.addTree(commit.getTree());
            tw.setRecursive(true);
            var rawChanges = DiffEntry.scan(tw);
            sendChanges(rawChanges.stream()
                    .map(it -> {
                        try {
                            return processDiff(it);
                        } catch (IOException e) {
                            throw new RuntimeException(e);
                        }
                    })
                    .flatMap(List::stream)
                    .collect(Collectors.toList())
            );
        }
    }

    Supplier<String> fileGetter(AbbreviatedObjectId abbreviatedObjectId) {
        return () -> {
            try {
                return new String(repository.open(abbreviatedObjectId.toObjectId()).getBytes());
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        };
    }

    List<Change> processDiff(DiffEntry diffEntry) throws IOException {
        return switch (diffEntry.getChangeType()) {
            case ADD -> List.of(new AddChange(System.currentTimeMillis(),
                    new FilePointer(new File(diffEntry.getNewPath()), 0),
                    new String(repository.open(diffEntry.getNewId().toObjectId()).getBytes()))
            );
            case MODIFY -> List.of(
                    new ModifyChange(System.currentTimeMillis(),
                            fileGetter(diffEntry.getOldId()),
                            fileGetter(diffEntry.getNewId()),
                            new File(diffEntry.getOldPath()),
                            new File(diffEntry.getNewPath())
                    ));
            case DELETE -> List.of(
                    new DeleteChange(System.currentTimeMillis(), new FilePointer(new File(diffEntry.getOldPath()), 0),
                            new String(repository.open(diffEntry.getOldId().toObjectId()).getBytes())));
            case RENAME -> List.of(
                    new RenameChange(System.currentTimeMillis(),
                            fileGetter(diffEntry.getOldId()),
                            fileGetter(diffEntry.getNewId()),
                            new File(diffEntry.getOldPath()),
                            new File(diffEntry.getNewPath())
                    ));
            case COPY -> List.of(
                    new CopyChange(System.currentTimeMillis(),
                            fileGetter(diffEntry.getOldId()),
                            fileGetter(diffEntry.getNewId()),
                            new File(diffEntry.getOldPath()),
                            new File(diffEntry.getNewPath())
                    ));
        };
    }

    private void parseFirstCommit(RevCommit first) {
        List<Change> changes = new ArrayList<>();
        try (TreeWalk treeWalk = new TreeWalk(repository)) {
            treeWalk.addTree(first.getTree());
            treeWalk.setRecursive(true);
            while (treeWalk.next()) {
                changes.add(new AddChange(System.currentTimeMillis(),
                        new FilePointer(new File(treeWalk.getPathString()), 0),
                        new String(repository.open(treeWalk.getObjectId(0)).getBytes()))
                );
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        sendChanges(changes);
    }
}

File: src/main/java/com/github/sudu/persistentidecaches/GitParser.kt
package com.github.sudu.persistentidecaches

import com.github.sudu.persistentidecaches.GitParser
import com.github.sudu.persistentidecaches.changes.*
import com.github.sudu.persistentidecaches.lmdb.maps.LmdbSha12Int
import com.github.sudu.persistentidecaches.records.FilePointer
import com.github.sudu.persistentidecaches.records.Revision
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.api.errors.GitAPIException
import org.eclipse.jgit.diff.DiffEntry
import org.eclipse.jgit.lib.AbbreviatedObjectId
import org.eclipse.jgit.lib.Constants
import org.eclipse.jgit.lib.ObjectId
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.revwalk.RevCommit
import org.eclipse.jgit.revwalk.RevWalk
import org.eclipse.jgit.treewalk.TreeWalk
import org.eclipse.jgit.treewalk.filter.AndTreeFilter
import org.eclipse.jgit.treewalk.filter.IndexDiffFilter
import org.eclipse.jgit.treewalk.filter.PathSuffixFilter
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.Path
import java.util.*
import java.util.function.Supplier
import java.util.stream.Collectors
import kotlin.math.min

class GitParser @JvmOverloads constructor(
    git: Git, private val indexesManager: IndexesManager, private val gitCommits2Revisions: LmdbSha12Int,
    private val commitsLimit: Int = Int.MAX_VALUE
) {
    private val repository: Repository = git.repository

    @Throws(IOException::class)
    fun parseAll() {
        val refs = repository.refDatabase.refs
        System.err.println("Parsing " + refs.size + " refs")
        var cnt = 0
        for (ref in refs) {
            parseOne(ref.objectId)
            System.err.println("Parsed " + ref.name + " " + (++cnt) + "/" + refs.size + " refs")
        }
    }

    @Throws(IOException::class)
    fun parseHead() {
        parseOne(repository.resolve(Constants.HEAD))
    }

    @get:Throws(IOException::class)
    val head: String
        get() {
            RevWalk(repository).use { walk ->
                return walk.parseCommit(repository.resolve(Constants.HEAD)).name
            }
        }

    private fun parseOne(head: ObjectId) {
        LOG.info("Parsing ref: " + head.name)
        try {
            RevWalk(repository).use { walk ->
                val commits: Deque<RevCommit> = ArrayDeque()
                var firstCommit: RevCommit? = null
                run {
                    walk.markStart(walk.parseCommit(head))
                    for (commit in walk) {
                        commits.add(commit)
                        if (gitCommits2Revisions.get(commit.name) != -1) {
                            firstCommit = commit
                            break
                        }
                    }
                }
                if (!commits.iterator().hasNext()) {
                    throw RuntimeException("Repository hasn't commits")
                }
                LOG.info(String.format("%d commits found to process", commits.size))

                if (firstCommit == null) {
                    indexesManager.checkout(Revision.NULL)
                    firstCommit = commits.removeLast()
                    parseFirstCommit(firstCommit)
                } else {
                    val rev = Revision(gitCommits2Revisions.get(firstCommit!!.name))
                    indexesManager.checkout(rev)
                }
                var prevCommit = firstCommit

                var commitsParsed = 0
                val totalCommits = min(commitsLimit.toDouble(), commits.size.toDouble()).toInt()
                while (commitsParsed < totalCommits) {
                    if (commitsParsed % 100 == 0) {
                        System.err.printf("Processed %d commits out of %d %n", commitsParsed, totalCommits)
                    }
                    val commit = commits.removeLast()
                    parseCommit(commit, prevCommit)
                    prevCommit = commit
                    commitsParsed++
                }
                System.err.println("Processed $totalCommits commits")
            }
        } catch (e: GitAPIException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }


    @Throws(IOException::class, GitAPIException::class)
    private fun parseCommit(commit: RevCommit, prevCommit: RevCommit?) {
        TreeWalk(repository).use { tw ->
            tw.addTree(prevCommit!!.tree)
            tw.addTree(commit.tree)
            tw.filter = IndexDiffFilter.ANY_DIFF
            tw.filter = AndTreeFilter.create(IndexDiffFilter.ANY_DIFF, PathSuffixFilter.create(".java"))
            tw.isRecursive = true
            val rawChanges = DiffEntry.scan(tw)
            sendChanges(rawChanges.stream()
                .map<List<Change>> { it: DiffEntry ->
                    try {
                        return@map processDiff(it)
                    } catch (e: IOException) {
                        throw RuntimeException(e)
                    }
                }
                .flatMap<Change> { obj: List<Change> -> obj.stream() }
                .collect(Collectors.toList()),
                commit
            )
        }
    }

    fun sendChanges(changes: List<Change>, commit: RevCommit) {
        changes.forEach {
            Objects.requireNonNull(it)
            if (it is FileChange) {
                indexesManager.fileCache.tryRegisterNewObj(it.place.file)
            } else if (it is FileHolderChange) {
                indexesManager.fileCache.tryRegisterNewObj(it.oldFileName)
                indexesManager.fileCache.tryRegisterNewObj(it.newFileName)
            }
        }
        val rev = gitCommits2Revisions.get(commit.name)
        if (rev == -1) {
            indexesManager.revisions.currentRevision = indexesManager.revisions.addRevision(
                indexesManager.revisions.currentRevision
            )
            gitCommits2Revisions.put(commit.name, indexesManager.revisions.currentRevision.revision)
            indexesManager.applyChanges(changes)
        } else {
            indexesManager.checkout(Revision(rev))
        }
    }


    fun fileGetter(abbreviatedObjectId: AbbreviatedObjectId): Supplier<String> {
        return Supplier {
            try {
                return@Supplier String(repository.open(abbreviatedObjectId.toObjectId()).bytes)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }

    @Throws(IOException::class)
    fun processDiff(diffEntry: DiffEntry): List<Change> {
        return when (diffEntry.changeType) {
            DiffEntry.ChangeType.ADD -> java.util.List.of<Change>(
                AddChange(
                    System.currentTimeMillis(),
                    FilePointer(Path.of(diffEntry.newPath), 0),
                    String(repository.open(diffEntry.newId.toObjectId()).bytes)
                )
            )

            DiffEntry.ChangeType.MODIFY -> java.util.List.of<Change>(
                ModifyChange(
                    System.currentTimeMillis(),
                    fileGetter(diffEntry.oldId),
                    fileGetter(diffEntry.newId),
                    Path.of(diffEntry.oldPath),
                    Path.of(diffEntry.newPath)
                )
            )

            DiffEntry.ChangeType.DELETE -> java.util.List.of<Change>(
                DeleteChange(
                    System.currentTimeMillis(),
                    FilePointer(Path.of(diffEntry.oldPath), 0),
                    String(repository.open(diffEntry.oldId.toObjectId()).bytes)
                )
            )

            DiffEntry.ChangeType.RENAME -> java.util.List.of<Change>(
                RenameChange(
                    System.currentTimeMillis(),
                    fileGetter(diffEntry.oldId),
                    fileGetter(diffEntry.newId),
                    Path.of(diffEntry.oldPath),
                    Path.of(diffEntry.newPath)
                )
            )

            DiffEntry.ChangeType.COPY -> java.util.List.of<Change>(
                CopyChange(
                    System.currentTimeMillis(),
                    fileGetter(diffEntry.oldId),
                    fileGetter(diffEntry.newId),
                    Path.of(diffEntry.oldPath),
                    Path.of(diffEntry.newPath)
                )
            )
        }
    }

    private fun parseFirstCommit(first: RevCommit?) {
        val rev = gitCommits2Revisions.get(first!!.name)
        if (rev != -1) {
            indexesManager.revisions.currentRevision = Revision(rev)
            return
        }
        val changes: MutableList<Change> = ArrayList()
        try {
            TreeWalk(repository).use { treeWalk ->
                treeWalk.addTree(first.tree)
                treeWalk.filter = AndTreeFilter.create(IndexDiffFilter.ANY_DIFF, PathSuffixFilter.create(".java"))
                treeWalk.isRecursive = true
                while (treeWalk.next()) {
                    changes.add(
                        AddChange(
                            System.currentTimeMillis(),
                            FilePointer(Path.of(treeWalk.pathString), 0),
                            String(repository.open(treeWalk.getObjectId(0)).bytes)
                        )
                    )
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        sendChanges(changes, first)
    }

    companion object {
        const val PARSE_ONLY_TREE: Boolean = false
        private val LOG: Logger = LoggerFactory.getLogger(GitParser::class.java)
    }
}